src/gpg/sign.py

```diff
@@ -1,5 +1,15 @@
 import sys
 import gpg
+import os
+
+# if autopin is declared (tests) use it
+try:
+    autopin = os.environ["autopin"]
+    sys.path.insert(0, autopin.rsplit("/", 1)[0])
+    exec('import autopin')
+    sys.path.pop(0)
```

Comment:
It's not clear what you want achieve here. If `autopin` variable is set then import autopin python module?

Reply:
This is required in `tests` when we use this to automatically enter the pin.

Reply:
No need for sorry. You're doing a great job :)

Reply:
BTW The intent here is to attach a autopin function to gpg.Context. This is required in tests. I am using environment variable autopin which contains path to the file autopin.py.

Reply:
> The intent here is to attach a autopin function to gpg.Context. This is required in tests.

I think that you are doing it the wrong way. This is not how gpg works. You should research more about how **pinentry** works in gpg.
Your `autopin.py` is not equivalent to `autopin.sh` that it has replaced. You can also keep `autopin.sh`, no need to rewrite it in python, as long as it does the job.

Reply:
As said on gpg devel list, gpgme does not respect gpg-conf.
(Without this tests are failing but that is mainly because autopin.sh is
not working first of all)
This way is similar to what the tests/examples on gpgme use.
But still I am working on correcting autopin.sh first. If it works it would
be better.

Reply:
As said on gpg devel list, gpgme does not respect gpg-conf.
(Without this tests are failing but that is mainly because autopin.sh is
not working first of all)
This way is similar to what the tests/examples on gpgme use.
But still I am working on correcting autopin.sh first. If it works it would
be better.

Reply:
Gpgme does launch an out of band prompt for pinentry by default. 
The other option it provides is to attach a callback function(this is what I have used).

If gpgme respects gpg-conf, the autopin.sh would work by default and will solve the problem. (On correcting autopin.sh)

Reply:
> gpgme does not respect gpg-conf 

`gpg.conf` maybe is deprecated on `gpg2` (this is not related to `gpgme`)
However we are using it on `gpg-agent.conf`, which is a different thing. If this configuration is deprecated too, then we should find where is the right place to do the configuration of pinentry.

Reply:
For `egpg` the tests that use `autopin.sh` work (t24 and t28).
The purpose of `autopin.sh` is to automatically enter the passphrase of the key, in order to automate testing steps. If it fails, one of the reasons can also be that the passphrase of the key is different from what `pinentry.sh` supplies.

Reply:
@dashohoxha as indicated by the logs above in the first message I feel they are giving false positives,  particularly please have a look at the following lines the passphrase is not getting changed. Can you confirm once again if they actually work. It is for egpg gnupg2.2 branch. The passphrase is not changing in my opinion. Further things work as the existing keys (for which the passphrase is changed) have no existing passphrase.

```
this is for egpg
[GNUPG:] PINENTRY_LAUNCHED 31354 tty 1.1.0 not a tty dumb -
gpg: key 18D1DA4D9E7A4FD0/18D1DA4D9E7A4FD0: error changing passphrase: End of file
[GNUPG:] PINENTRY_LAUNCHED 31355 tty 1.1.0 not a tty dumb -
gpg: key 18D1DA4D9E7A4FD0/12A2B2669B636DD4: error changing passphrase: End of file
gpg: error changing the passphrase for '18D1DA4D9E7A4FD0': End of file
[GNUPG:] ERROR keyedit.passwd 83902463
```

Reply:
> The purpose of autopin.sh is to automatically enter the passphrase of the key, in order to automate testing steps. If it fails, one of the reasons can also be that the passphrase of the key is different from what pinentry.sh supplies.
Ok I'll look into this.

Reply:
- Dashamir, I looked more about gpg-agent have been able to get the autopin script working partially ( on **seal** and **sign** ). The problem was-
  - the GNUPGHOME is not being set up correctly in setup_autopin

- Strangely autopin is not working for `egpg key pass` i.e for changing passphrase, with same configuration and settings `pinentry` is being launched instead. 
I tried this by
  -  preserving the trash directory for t24 and setting EGPG_DIR and GNUPGHOME to those of trash dir. 
  - ran `egpg key pass` and pinentry was launched. set new key to `0123456789` as in the tests
  - ran `egpg sign ... it is automatically using autopin.sh (i have confirmed this by adding echo to file statements in autopin.sh)
  - ran `egpg key pass` again this time I got a pinentry prompt for old passphrase. 

This I believe is strange as it is launching different pinentry scripts for new password and signing respectively.

Need your suggestions on how to debug this further. I seem to be stuck. Also do you remember it working earlier like in egpg-2.0 or 2.1

Reply:
> do you remember it working earlier like in egpg-2.0 or 2.1

I think it has been working, but I am not sure, maybe the test was working for the wrong reasons.

> This I believe is strange as it is launching different pinentry scripts for new password and signing respectively.

I think it should be the same, unless the configuration on `gpg-agent.conf` has changed.

> the GNUPGHOME is not being set up correctly in setup_autopin

You can check it by adding a line like `echo $GNUPGHOME` at the end of `setup_autopin()` (and using the option `-d` on `./run.sh -d t24-key-pass.t`)

> Strangely autopin is not working for egpg key pass i.e for changing passphrase, with same configuration and settings pinentry-tty is being launched instead.

You are right. The configuration `gpg-agent.conf` that is created on `setup_autopin()` somehow is ignored (or `pinentry-program` is ignored), and `pinentry-tty` is used instead. I don't know why and I don't know how to debug it further.

I have tried to debug by adding these lines on `setup_autopin()`:
```
    set -x
    cat <<EOF >> "$GNUPGHOME"/gpg-agent.conf
debug-level guru
debug-all
debug-pinentry
EOF
    cat "$GNUPGHOME"/gpg-agent.conf
    gpgconf
    gpgconf --list-options gpg-agent
    gpg-connect-agent reloadagent /bye
    gpgconf --launch gpg-agent
    set +x
```

But it doesn't help. I am stuck too. Sorry, I cannot help you on this.

So, if you are tired, you can skip this for the time being and add it in the ToDo list (of things that should be checked later). It would be nice to have this automated test that checks password change, and if we manage to do it right we can learn a lot about how `gpg` works.
But it is not an absolute requirement, we can test password changing manually as well.